### PR TITLE
ROX-25822: Query nvd cvss for vuln report

### PR DESCRIPTION
**Pull request by ksurabhi91**
This PR adds nvd cvss field to vuln report
Tested these changes by manually dwnloading report when nvdcvss filter is set to truwe and when filter is set to false
[RHACS_Vulnerability_Report_Test_09_October_2024_false.csv](https://github.com/user-attachments/files/17318839/RHACS_Vulnerability_Report_Test_09_October_2024_false.csv)
[RHACS_Vulnerability_Report_Test_08_October_2024 2_true.csv](https://github.com/user-attachments/files/17318840/RHACS_Vulnerability_Report_Test_08_October_2024.2_true.csv)

### User-facing documentation

- [ ] CHANGELOG is updated **OR** update is not needed
- [ ] [documentation PR](https://spaces.redhat.com/display/StackRox/Submitting+a+User+Documentation+Pull+Request) is created and is linked above **OR** is not needed

### Testing and quality

- [ ] the change is production ready: the change is GA or otherwise the functionality is gated by a feature flag
- [ ] CI results are inspected

#### Automated testing

- [ ] added unit tests
- [ ] added e2e tests
- [ ] added regression tests
- [ ] added compatibility tests
- [ ] modified existing tests

#### How I validated my change

change me!

### CODE REVIEW

File: central/reports/common/report_formatter.go [hunk 1]
```diff
@@ -143,6 +152,7 @@ func Format(deployedImagesResults []DeployedImagesResult, watchedImagesResults [
 						v.FixedByVersion,
 						strings.ToTitle(stringutils.GetUpTo(v.Severity, "_")),
 						strconv.FormatFloat(v.Cvss, 'f', 2, 64),
+						strconv.FormatFloat(v.Nvdcvss, 'f', 2, 64),
```
Comment (dashrews78) [inline central/reports/common/report_formatter.go:155]: Don't we need to only add this when `includeNvd` is true?

More general question, could we just always include the NVD and if it is empty then we just don't have it?

Comment (ksurabhi91) [reply central/reports/common/report_formatter.go:155]: @dashrews78  this ads filed for select query but nvd cvss column is only added when flag is set to true 
https://github.com/stackrox/stackrox/pull/12870/files#diff-ac5eae65a138982ce2c2546e12b29d9f5d40e084f68acefb7d46dbd6419eb820R95

Comment (dashrews78) [reply central/reports/common/report_formatter.go:155]: That doesn't look like what the code is doing to me.  I'm not familiar with reports, but this looks like it is writing that field to the CSV for watched images.  Seems to me like you would need logic similar to what you have at line 124.

File: central/reports/common/report_formatter.go [hunk 2]
```diff
@@ -115,7 +120,11 @@ func Format(deployedImagesResults []DeployedImagesResult, watchedImagesResults [
 							strconv.FormatFloat(v.Cvss, 'f', 2, 64),
 							discoveredTs,
 							v.Link,
-						})
+						}
+						if includeNvd == true {
+							row = append(row, strconv.FormatFloat(v.Nvdcvss, 'f', 2, 64))
```
Comment (charmik-redhat) [inline central/reports/common/report_formatter.go:125]: Can you add a func `AppendToValue(val Value, field string)` or something to `pkg/csv/writer.go` . That we we don't have to know that Value is a string slice here.

File: central/reports/common/report_formatter.go (hunk 1 above)
Comment (charmik-redhat) [reply central/reports/common/report_formatter.go:155]: Similar to how you did for deployed images, you might only want to add this if `includeNvd` is true

File: central/reports/scheduler/schedule.go [hunk 3]
```diff
@@ -278,7 +278,8 @@ func (s *scheduler) sendReportResults(req *ReportRequest) error {
 		return err
 	}
 	// Format results into CSV
-	zippedCSVResult, err := common.Format(reportData, nil, "")
+	includeNvd := rc.GetVulnReportFilters().GetIncludeNvdCvss()
+	zippedCSVResult, err := common.Format(reportData, nil, "", includeNvd)
```
Comment (charmik-redhat) [inline central/reports/scheduler/schedule.go:282]: You can just send `includeNvd` as `false` here. This is v1 report implementation. I will create a ticket to remove this v1 reporting code.

File: central/reports/scheduler/v2/reportgenerator/report_gen_impl.go
Comment (charmik-redhat) [inline central/reports/scheduler/v2/reportgenerator/report_gen_impl.go:1]: You would need to update the graphQL queries to get `NvdCVSS` too. The queries are defined in `central/reports/scheduler/v2/reportgenerator/consts.go`

You could update the query based in whether `includeNvdCvss` is true or not. If that is not straight forward, you can also just change the consts to always get the field.

**Review by charmik-redhat — COMMENTED** [review]
Need to update the graphQL queries for deployed and watched images to get NvdCvss.

File: central/reports/common/report_formatter.go [hunk 4]
```diff
@@ -89,8 +90,12 @@ type ZippedCSVResult struct {
 
 // Format takes in the results of vuln report query, converts to CSV and returns zipped CSV data and
 // a flag if the report is empty or not. For v1 config pass an empty string.
-func Format(deployedImagesResults []DeployedImagesResult, watchedImagesResults []WatchedImagesResult, configName string) (*ZippedCSVResult, error) {
-	csvWriter := csv.NewGenericWriter(csvHeader, true)
+func Format(deployedImagesResults []DeployedImagesResult, watchedImagesResults []WatchedImagesResult, configName string, includeNvd bool) (*ZippedCSVResult, error) {
+	csvHeaderRep := csvHeader
```
Comment (charmik-redhat) [inline central/reports/common/report_formatter.go:94]: Not sure how go does it. Does this assignment create a copy of `csvHeader` ? Or would updates to `csvHeaderRep` still get applied to it?

File: central/reports/common/report_formatter.go (hunk 1 above)
Comment (charmik-redhat) [reply central/reports/common/report_formatter.go:156]: +1

File: pkg/csv/writer.go [hunk 5]
```diff
@@ -39,6 +39,11 @@ func (c *GenericWriter) AddValue(value Value) {
 	c.values = append(c.values, value)
 }
 
+// Append to value
+func (c *GenericWriter) AppendValue(value Value, field string) {
```
Comment (charmik-redhat) [inline pkg/csv/writer.go:43]: nit: Rename this to `AppendToValue`  to avoid confusion with existing `AddValue` function.

**Review by charmik-redhat — COMMENTED** [review]
Can you run a manual test with some watched images too?

File: central/reports/common/report_formatter.go (hunk 4 above)
Comment (ksurabhi91) [reply central/reports/common/report_formatter.go:94]: I have tested this. Go will create a new copy of csvheader

**Review by charmik-redhat — APPROVED** [review]

**Pull request merged**